candle-nn/src/encoding.rs

```diff
@@ -0,0 +1,555 @@
+///
+/// This method will bail if the index value is less than -1.
+///
+/// This method will bail if the index value is greater than or equal to the depth value.
+///
+/// This method will bail if the input data type is `f64`.
+///
+/// # API Design
+///
+/// The api design for this method is loosely based on the [TensorFlow One-Hot](https://www.tensorflow.org/api_docs/python/tf/one_hot) method.
+pub fn one_hot<D: WithDType>(
+    indices: Tensor,
+    depth: usize,
+    on_value: Option<D>,
+    off_value: Option<D>,
+) -> Result<Tensor> {
+    let on_value = on_value.unwrap_or(D::from_f64(1.));
+    let off_value = off_value.unwrap_or(D::from_f64(0.));
+
+    let dtype = indices.dtype();
+    let rank = indices.rank();
+
+    match rank {
+        0 => {
+            let mut v = vec![off_value; depth];
+            let idx = 0;
+            match dtype {
+                DType::U8 => {
+                    let vi = indices.to_vec0::<u8>()?;
+                    set_uint_value(vi as usize, idx, depth, &mut v, on_value)?;
+                }
+                DType::U32 => {
+                    let vi = indices.to_vec0::<u32>()?;
+                    set_uint_value(vi as usize, idx, depth, &mut v, on_value)?;
+                }
+                DType::I64 => {
+                    let vi = indices.to_vec0::<i64>()?;
+                    set_int_value(vi, idx, depth, &mut v, on_value)?;
+                }
+                d => unsupported_dtype(d)?,
+            };
+
+            Tensor::new(v, indices.device())
+        }
+        1 => {
+            let dim1 = indices.dims1()?;
+            let mut v = vec![off_value; depth * dim1];
+
+            match dtype {
+                DType::U8 => {
+                    let index = indices.to_vec1::<u8>()?;
+                    for i in 0..dim1 {
+                        set_uint_value(index[i] as usize, i * depth, depth, &mut v, on_value)?;
+                    }
+                }
+                DType::U32 => {
+                    let index = indices.to_vec1::<u32>()?;
+                    for i in 0..dim1 {
+                        set_uint_value(index[i] as usize, i * depth, depth, &mut v, on_value)?;
+                    }
+                }
+                DType::I64 => {
+                    let index = indices.to_vec1::<i64>()?;
+                    for i in 0..dim1 {
+                        set_int_value(index[i], i * depth, depth, &mut v, on_value)?;
+                    }
+                }
+                d => unsupported_dtype(d)?,
+            };
+
+            Tensor::new(v, indices.device())?.reshape(&[dim1, depth])
+        }
+        2 => {
+            let (dim1, dim2) = indices.dims2()?;
+            let mut v = vec![off_value; depth * dim1 * dim2];
+
+            let idx = |i: usize, j: usize, depth: usize, dim2: usize| -> usize {
+                i * depth * dim2 + j * depth
+            };
+
+            let iter = (0..dim1).flat_map(|i| (0..dim2).map(move |j| (i, j)));
```

Comment:
Is the `move` really necessary here? I would have thought that we don't care as `usize` is `Copy` (I think the same comment applies at other places).